Feature: add DFT-1/2 and shell DFT-1/2, only support PW esolvers_ks_pw. (Useful information to support DFT-1/2 method with pw basis set)

Wuming-HUST opened the pull request:
Feature: add DFT-1/2 and shell DFT-1/2, currently only support PW esolver_ks_pw.

* Added Sep, Sep_Cell, and VSep to organize the self-energy potential of DFT-1/2

* Added a new effective potential pot_sep for calculating the self-energy potential

* Added initialization of the self-energy potential in the esolver_ks_pw control flow

* Added the keyword SEP_FILES in the STRU file for reading self-energy potential files

* Added the dfthalf_type keyword in INPUT to enable DFT-1/2 and shell DFT-1/2

### Reminder
- [x] Have you added adequate unit tests and/or case tests for your pull request?
- [x] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Unit Tests and/or Case Tests for my changes
- Added unit test for Sep and Sep_Cell.

### What's changed?
- I have additionally added the reading and processing of the self-consistent potential, which is currently only available in esolver_ks_pw. 
- The calculation of pot_sep is similar to that of pot_local, completed only during before_scf, having almost no impact on overall performance.
- The STRU file has been enhanced with a keyword for reading self-energy potential files (.sep), and the format has been described.

Wuming-HUST committed:
Feature: add DFT-1/2 and shell DFT-1/2, currently only support PW esolver_ks_pw.

* Added Sep, Sep_Cell, and VSep to organize the self-energy potential of
DFT-1/2

* Added a new effective potential pot_sep for calculating the
self-energy potential

* Added initialization of the self-energy potential in the esolver_ks_pw
control flow

* Added the keyword SEP_FILES in the STRU file for reading self-energy
potential files

* Added the dfthalf_type keyword in INPUT to enable DFT-1/2 and shell
DFT-1/2

dependabot[bot] committed:
Build(deps): Bump actions/setup-python from 5 to 6 (#6492)

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 5 to 6.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Critsium-xy committed:
[Refactor] Move hardware initializer out from esolver code (#6494)

* Move hardware initializer out from esolver

* Remove useless codes

* Remove finalize code out

mohanchen commented:
It seems that this PR breaks some tests for other functionalities. Besides, we should try avoid using GlobalC in the code.

wangtianxiang committed:
Feature: support NVTX profiling via timer_enable_nvtx flag (#6495)

* Feature: support NVTX profiling via timer_enable_nvtx flag
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Add timer_enable_nvtx section in markdown
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Fix: Use __USE_NVTX macro to avoid NVTX linking errors in tests.
Clarify in docs that timer_enable_nvtx parameter only takes effect on CUDA platforms.
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

wangtianxiang committed:
Perf: Optimize Davidson by fusing operators, offloading CPU computation to GPU, and reducing memory transfers (#6493)

* Perf: Optimize Diago_DavSubspace with GPU operators by adding and fusing custom kernels.
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Perf: reduce memory allocation and copy in Diago_DavSubspace::diag_zhegvx
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Perf: Replace loop-based 2D copy and memset with memcpy_2d_op, memset_2d_op
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Perf: use warp reduce instead of shared memory for better efficiency
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Fix compilation error
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

wangtianxiang committed:
Fix: resolve compile error with USE_ELPA=OFF + BUILD_TESTING=ON and switch to nvtx3 headers when CUDA_VERSION >= 12090 (#6497)

* Fix: switch to nvtx3 headers when CUDA_VERSION >= 12090
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

* Fix: resolve compile error with USE_ELPA=OFF + BUILD_TESTING=ON
Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

Wuming-HUST committed:
Fix: Compilation error in DeepKS unit tests after adding DFT-1/2

Critsium-xy committed:
Fix dsp compilation problem (#6499)

wangtianxiang committed:
Fix: Fix crash in Debug build with multi-GPU due to forced cudaSetDevice(0) (#6498)

Signed-off-by：Tianxiang Wang<tianxiang.wang@metax-tech.com>, Contributed under MetaX Integrated Circuits (Shanghai) Co., Ltd.

zgn-26714 committed:
Removed the temporary variable DMRGint_full when transitioning from 2D block parallelism to serial in Hcontainer(develop) (#6489)

* delete tem Hcontainer to reduce memory usage

* simplify the compute code

* change DM2D_tmp to dm2d_tmp, use vector instead of new

YuLiu98 commented:
I have some questions:
1. What's the difference between `C_ca_50.sep` and `F_pbe_50.sep`, could you remove one of them to make the package smaller?
2. It may be better to merge `VSep_in_PW.*` into `pot_sep.*`, just like `efield.*`.
3. We do not recommend the usage of `GlobalC `.

ErjieWu committed:
Update version to 3.9.0.14 (#6504)

Wuming-HUST committed:
Fix: Add the additional files to Makefile.Objects

mohanchen commented on source/source_estate/module_pot/pot_sep.h:
```diff
@@ -0,0 +1,46 @@
+    // }
+    PotSep(const ModuleBase::ComplexMatrix* sf_in, const ModulePW::PW_Basis* rho_basis_in) : sf_(sf_in)
+    {
+        assert(GlobalC::vsep_cell.vsep_form.nr == this->sf_->nr);
```
maybe use can you vsep_cell.vsep_form.nr as an input parameter, then you can avoid using GlobalC::vsep_cell

mohanchen commented on source/source_estate/module_pot/pot_sep.cpp:
```diff
@@ -0,0 +1,25 @@
+
+    for (int ir = 0; ir < this->rho_basis_->nrxx; ++ir)
+    {
+        vl_pseudo[ir] += GlobalC::vsep_cell.vsep_r[ir];
```
maybe you can use GlobalC::vsep_cell.vsep_r as input instead of using GlobalC

mohanchen commented on source/source_esolver/esolver_ks_pw.cpp:
```diff
@@ -262,6 +268,13 @@ void ESolver_KS_PW<T, Device>::before_scf(UnitCell& ucell, const int istep)
+    //! 4.5) DFT-1/2 calculations, sep potential need to generate before effective potential calculation
+    //----------------------------------------------------------
+    if (PARAM.inp.dfthalf_type > 0) {
+        GlobalC::vsep_cell.generate_vsep_r(this->pw_rhod[0], this->sf.strucFac);
```
maybe you can let vsep_cell become a member of esolver_ks_pw class

mohanchen commented on source/source_cell/unitcell.cpp:
```diff
@@ -329,11 +345,13 @@ void UnitCell::setup_cell(const std::string& fn, std::ofstream& log)
     //===================================
     this->set_iat2itia();
 
+    GlobalC::sep_cell.set_omega(this->omega, this->tpiba2);
```
maybe you can move GlobalC::sep_cell.set_omega to esolver instead of inside unitcell, then you can avoid using sep_cell as a GlobalC

mohanchen commented on source/source_cell/unitcell.cpp:
```diff
@@ -245,6 +249,13 @@ void UnitCell::setup_cell(const std::string& fn, std::ofstream& log)
+            // readl sep potential, currently using the pseudopotential folder (pseudo_dir in INPUT)
+            //==========================
+            if (PARAM.inp.dfthalf_type > 0) {
+                GlobalC::sep_cell.init(this->ntype);
```
maybe sep_cell can be a member of unitcell class?

Wuming-HUST committed:
Refactor: Remove the GlobalC from sep_cell and vsep_cell

* Removed GlobalC::sep_cell and GlobalC::vsep_cell from GlobalC

* Integrated sep_cell into UnitCell

* Integrated vsep_cell into esolver_ks_pw

* Added empty constructors and destructors for Sep_Pot and Sep_Cell to
facilitate unit testing compilation

mohanchen committed:
Merge branch 'develop' into feature/dft-half

mohanchen commented:
#6470 solved
